### PR TITLE
core: Add basic support for Tab key behavior

### PR DESCRIPTION
**Pull request by kjarosh**
Related to #5443. Improves coverage for AVM1: #243, #263, #280.

The Tab key is used to cycle through focusable elements in stage. It supports two tab orderings: automatic and custom. This patch adds basic support for this behavior and both modes.

Things implemented:
* Tab key now cycles through the tab ordering and changes the focus.
* Tab key now works on desktop (instead of cycling through interactive GUI elements).
* Automatic tab ordering.
* Custom tab ordering.
* `TextField.tabEnabled` in AVM1.
* `TextField.tabIndex` in AVM1.
* `Button.tabEnabled` in AVM1.
* `Button.tabIndex` in AVM1.
* `MovieClip.tabEnabled` in AVM1.
* `MovieClip.tabIndex` in AVM1.

Things known to be missing/broken yet:
* `MovieClip.tabChildren` in AVM1 is not implemented.
* No AVM2 support yet.
* There's no visual highlight of the tabbed element.
* The order in automatic tab ordering is not fine-tuned.

### CODE REVIEW

File: desktop/src/gui/controller.rs [hunk 1]
```diff
@@ -190,14 +192,28 @@ impl GuiController {
...
+            &event,
+            WindowEvent::KeyboardInput {
+                event: winit::event::KeyEvent {
+                    logical_key: Key::Named(NamedKey::Tab),
+                    ..
+                },
+                ..
+            }
+        ) {
+            // Prevent egui from consuming the Tab key.
```
Comment (Lord-McSweeney) [inline desktop/src/gui/controller.rs:213]: On #13426, it was stated that
> The TAB key is used to access the inventory in the game, but in the Ruffle desktop player, TAB moves through Ruffle's own menus and doesn't forward the key to the game. Fortunately, you can also open the inventory by clicking the backpack with the mouse...

Does this change fix that?

Comment (kjarosh) [reply desktop/src/gui/controller.rs:213]: That is true, it should be fixed now.

**Review by adrian17 — APPROVED** [review]
Approved, assuming the nits are fixed and that weird behavior I found on discord is documented (maybe a failing test?)

File: core/src/display_object/avm1_button.rs [hunk 2]
```diff
@@ -46,6 +46,7 @@ pub struct Avm1ButtonData<'gc> {
     object: Lock<Option<Object<'gc>>>,
     initialized: Cell<bool>,
     has_focus: Cell<bool>,
+    tab_index: Cell<Option<i32>>,
```
[the remark this answers is not in the document]
Comment (kjarosh) [reply core/src/display_object/avm1_button.rs:51]: Done, added a TODO for each one

File: core/src/display_object.rs [hunk 3]
```diff
@@ -1890,11 +1890,23 @@ pub trait TDisplayObject<'gc>:
...
 
+    /// Whether this object is included in tab ordering.
+    fn is_tab_enabled(&self, _context: &mut UpdateContext<'_, 'gc>) -> bool {
+        false
+    }
+
+    /// Used to customize tab ordering.
+    /// When not `None`, a custom ordering is used, and
+    /// objects are ordered according to this value.
+    fn get_tab_index(&self) -> Option<i64> {
```
[the remark this answers is not in the document]
Comment (kjarosh) [reply core/src/display_object.rs]: Fixed, removed the prefix from all getters

File: core/src/display_object/movie_clip.rs [hunk 4]
```diff
@@ -2584,6 +2562,14 @@ impl<'gc> MovieClip<'gc> {
             }
         }
     }
+
+    pub fn get_tab_index_value(&self) -> Option<i32> {
+        self.0.read().tab_index
+    }
+
+    pub fn set_tab_index_value(&self, context: &mut UpdateContext<'_, 'gc>, value: Option<i32>) {
```
[the remark this answers is not in the document]
Comment (kjarosh) [reply core/src/display_object/movie_clip.rs:2574]: Removed the prefix, but left the `_value` suffix due to the i32/u32 thingy and a potential name conflict with `tab_index`

**Comment by kjarosh** [comment]
Added a test `tab_ordering_properties_tab_index_edge_case` which documents the weird behavior from Discord.

**Pull request merged**